gemini/reasoning-engine/tutorial_graph_rag.ipynb

```diff
@@ -0,0 +1,1707 @@
+    "                                             spanner_database_id=self.spanner_database_id, embedding_model_name=self.embedding_model_name)\n",
+    "\n",
+    "        def format_docs(docs):\n",
+    "          return \"\\n\\n\".join(doc.page_content for doc in docs)\n",
+    "\n",
+    "        chain = (\n",
+    "            {\"context\": retriever | format_docs , \"question\": RunnablePassthrough()}\n",
+    "            | qa_prompt\n",
+    "            | llm\n",
+    "            | StrOutputParser()\n",
+    "        )\n",
+    "        return chain\n",
+    "\n",
+    "    def set_up(self):\n",
+    "\n",
+    "        # init the vertexai package\n",
+    "        vertexai.init(project=self.gcp_project_id, location=self.region)\n",
+    "        llm = ChatVertexAI(\n",
+    "            model_name=self.model_name,\n",
+    "            max_output_tokens=self.max_output_tokens,\n",
+    "            max_input_tokens=32000,\n",
+    "            temperature=self.temperature,\n",
+    "            top_p=self.top_p,\n",
+    "            top_k=self.top_k,\n",
+    "            project = self.gcp_project_id,\n",
+    "            location = self.region,\n",
+    "            # convert_system_message_to_human=True,\n",
+    "            response_validation=False,\n",
+    "            verbose=True\n",
+    "        )\n",
+    "\n",
+    "        self.qa_chain = self.configure_qa_rag_chain(llm)\n",
+    "\n",
+    "    def query(self, query):\n",
+    "        return self.qa_chain.invoke(query)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "B3o2czmN8Goo"
+   },
+   "source": [
+    "We will create the chain and test it out."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 26,
+   "metadata": {
+    "colab": {
+     "base_uri": "https://localhost:8080/"
+    },
+    "id": "QhxJy4omzP9J",
+    "outputId": "bb5c800c-9424-4637-e0a2-18e9d755b0d9",
+    "tags": []
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Maria Montessori.\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "lc = LangchainGraphRag(MODEL_NAME, EMBEDDING_MODEL_NAME, GCP_PROJECT_ID, REGION, SPANNER_INSTANCE_ID, SPANNER_DATABASE_ID)\n",
+    "lc.set_up()\n",
+    "\n",
+    "response = lc.query('Who influenced Larry Page?')\n",
+    "print(response)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "50jswr0X8cOK"
+   },
+   "source": [
+    "### Deploy to Vertex AI Reasoning Engine\n",
+    "\n",
+    "Great! The chain is working and return the answers from the orginal question. Now let's host this Langchain code on Vertex AI Reasoning Engine so we can call it as an API. First we will reorganize the code blocks above to make it work with the Vertex AI Reasoning Engine. Be sure to update the code with your Project ID and Region"
```

Comment:
We should have users define all constants in a cell near the top (like you have), then reuse that throughout to avoid having to re-define these in multiple places throughout the notebook.